File: engine/language_client_python/python_src/baml_py/stream.py
"""
BAML Stream with cancellation support for Python
"""

import asyncio
from typing import TypeVar, Generic, Optional, Callable, Iterator, AsyncIterator
import threading

T = TypeVar('T')
PartialT = TypeVar('PartialT')
FinalT = TypeVar('FinalT')


class BamlStream(Generic[PartialT, FinalT]):
    """
    A BAML stream that supports cancellation.
    
    This stream can be cancelled to stop ongoing HTTP requests to LLM providers,
    preventing wasted resources and billing charges.
    """
    
    def __init__(
        self,
        ffi_stream,
        partial_coerce: Callable[[any], PartialT],
        final_coerce: Callable[[any], FinalT],
        ctx_manager,
    ):
        self._ffi_stream = ffi_stream
        self._partial_coerce = partial_coerce
        self._final_coerce = final_coerce
        self._ctx_manager = ctx_manager
        self._cancelled = False
        self._final_result = None
        
    def cancel(self) -> None:
        """
        Cancel the stream processing.
        
        This will:
        1. Cancel the Rust-level stream
        2. Cancel ongoing HTTP requests to LLM providers
        3. Stop consuming network bandwidth and API quota
        4. Clean up resources
        """
        if not self._cancelled:
            self._cancelled = True
            self._ffi_stream.cancel()
    
    def is_cancelled(self) -> bool:
        """Check if the stream has been cancelled."""
        return self._cancelled or self._ffi_stream.is_cancelled()
    
    async def get_final_response(self) -> FinalT:
        """
        Get the final response from the stream.
        
        This will wait for the stream to complete and return the final result.
        If the stream is cancelled, this will raise an exception.
        """
        if self._final_result is not None:
            return self._final_result
            
        try:
            result = await self._ffi_stream.done(self._ctx_manager)
            final_result = self._final_coerce(result.parsed())
            self._final_result = final_result
            return final_result
        except Exception as e:
            if self.is_cancelled():
                raise RuntimeError("Stream was cancelled") from e
            raise
    
    def __aiter__(self) -> AsyncIterator[PartialT]:
        """Async iterator support for streaming partial results."""
        return self._async_iter()
    
    async def _async_iter(self) -> AsyncIterator[PartialT]:
        """Internal async iterator implementation."""
        partial_results = []
        
        def on_event(result):
            if not self.is_cancelled():
                try:
                    partial = self._partial_coerce(result.parsed())
                    partial_results.append(partial)
                except Exception as e:
                    # Log error but continue streaming
                    print(f"Error processing partial result: {e}")
        
        # Set up event handler
        self._ffi_stream.on_event(on_event)
        
        # Start the stream processing
        stream_task = asyncio.create_task(self.get_final_response())
        
        try:
            # Yield partial results as they come in
            last_index = 0
            while not stream_task.done() and not self.is_cancelled():
                # Yield any new partial results
                while last_index < len(partial_results):
                    yield partial_results[last_index]
                    last_index += 1
                
                # Small delay to avoid busy waiting
                await asyncio.sleep(0.01)
            
            # Yield any remaining partial results
            while last_index < len(partial_results):
                yield partial_results[last_index]
                last_index += 1
                
            # Wait for final result (or cancellation)
            await stream_task
            
        except Exception as e:
            if self.is_cancelled():
                raise RuntimeError("Stream was cancelled") from e
            raise
        finally:
            # Clean up event handler
            self._ffi_stream.on_event(None)


class BamlSyncStream(Generic[PartialT, FinalT]):
    """
    Synchronous version of BamlStream with cancellation support.
    """
    
    def __init__(
        self,
        ffi_stream,
        partial_coerce: Callable[[any], PartialT],
        final_coerce: Callable[[any], FinalT],
        ctx_manager,
    ):
        self._ffi_stream = ffi_stream
        self._partial_coerce = partial_coerce
        self._final_coerce = final_coerce
        self._ctx_manager = ctx_manager
        self._cancelled = False
        self._final_result = None
        
    def cancel(self) -> None:
        """
        Cancel the stream processing.
        
        This will:
        1. Cancel the Rust-level stream
        2. Cancel ongoing HTTP requests to LLM providers
        3. Stop consuming network bandwidth and API quota
        4. Clean up resources
        """
        if not self._cancelled:
            self._cancelled = True
            self._ffi_stream.cancel()
    
    def is_cancelled(self) -> bool:
        """Check if the stream has been cancelled."""
        return self._cancelled or self._ffi_stream.is_cancelled()
    
    def get_final_response(self) -> FinalT:
        """
        Get the final response from the stream.
        
        This will block until the stream completes and return the final result.
        If the stream is cancelled, this will raise an exception.
        """
        if self._final_result is not None:
            return self._final_result
            
        try:
            result = self._ffi_stream.done(self._ctx_manager)
            final_result = self._final_coerce(result.parsed())
            self._final_result = final_result
            return final_result
        except Exception as e:
            if self.is_cancelled():
                raise RuntimeError("Stream was cancelled") from e
            raise
    
    def __iter__(self) -> Iterator[PartialT]:
        """Iterator support for streaming partial results."""
        return self._sync_iter()
    
    def _sync_iter(self) -> Iterator[PartialT]:
        """Internal sync iterator implementation."""
        partial_results = []
        stream_done = threading.Event()
        
        def on_event(result):
            if not self.is_cancelled():
                try:
                    partial = self._partial_coerce(result.parsed())
                    partial_results.append(partial)
                except Exception as e:
                    # Log error but continue streaming
                    print(f"Error processing partial result: {e}")
        
        # Set up event handler
        self._ffi_stream.on_event(on_event)
        
        # Start stream processing in background thread
        def run_stream():
            try:
                self.get_final_response()
            except Exception:
                pass  # Error will be handled when get_final_response is called again
            finally:
                stream_done.set()
        
        stream_thread = threading.Thread(target=run_stream)
        stream_thread.start()
        
        try:
            # Yield partial results as they come in
            last_index = 0
            while not stream_done.is_set() and not self.is_cancelled():
                # Yield any new partial results
                while last_index < len(partial_results):
                    yield partial_results[last_index]
                    last_index += 1
                
                # Small delay to avoid busy waiting
                stream_done.wait(0.01)
            
            # Yield any remaining partial results
            while last_index < len(partial_results):
                yield partial_results[last_index]
                last_index += 1
                
        finally:
            # Clean up
            self._ffi_stream.on_event(None)
            stream_thread.join(timeout=1.0)  # Give it a second to clean up

File: engine/language_client_ruby/lib/stream.rb
require "sorbet-runtime"

module Baml
  class StreamState < T::Struct
    extend T::Sig
    extend T::Generic

    Value = type_member

    const :value, Value
    const :state, Symbol

    def initialize(props)
      super(value: props[:value], state: props[:state])
    end
  end

  class BamlStream
    extend T::Sig
    extend T::Generic

    include Enumerable

    PartialType = type_member
    FinalType = type_member

    def initialize(
      ffi_stream:,
      ctx_manager:
    )
      @ffi_stream = ffi_stream
      @ctx_manager = ctx_manager
      @final_response = nil
      @cancelled = false
      @mutex = Mutex.new
    end

    # Cancel the stream processing.
    # This will:
    # 1. Cancel the Rust-level stream
    # 2. Cancel ongoing HTTP requests to LLM providers
    # 3. Stop consuming network bandwidth and API quota
    # 4. Clean up resources
    sig { void }
    def cancel
      @mutex.synchronize do
        return if @cancelled
        @cancelled = true
        
        # Call the FFI stream's cancel method if it exists
        @ffi_stream.cancel if @ffi_stream.respond_to?(:cancel)
      end
    end

    # Check if the stream has been cancelled
    sig { returns(T::Boolean) }
    def cancelled?
      @mutex.synchronize { @cancelled } || 
        (@ffi_stream.respond_to?(:is_cancelled) && @ffi_stream.is_cancelled)
    end

    # Get the final response from the stream.
    # This will block until the stream completes and return the final result.
    # If the stream is cancelled, this will raise an exception.
    sig { returns(FinalType) }
    def get_final_response
      @mutex.synchronize do
        return @final_response if @final_response
      end

      raise "Stream was cancelled" if cancelled?

      begin
        result = @ffi_stream.done(@ctx_manager)
        final_result = coerce_final(result.parsed)
        
        @mutex.synchronize do
          @final_response = final_result
        end
        
        final_result
      rescue => e
        raise "Stream was cancelled: #{e.message}" if cancelled?
        raise
      end
    end

    # Iterate over partial results as they become available
    sig { params(block: T.proc.params(arg0: PartialType).void).void }
    def each(&block)
      return enum_for(:each) unless block_given?

      partial_results = []
      stream_done = false
      
      # Set up event handler for partial results
      @ffi_stream.on_event(proc do |result|
        next if cancelled?
        
        begin
          partial = coerce_partial(result.parsed)
          partial_results << partial
        rescue => e
          # Log error but continue streaming
          puts "Error processing partial result: #{e.message}"
        end
      end)

      # Start stream processing in background thread
      stream_thread = Thread.new do
        begin
          get_final_response
        rescue => e
          # Error will be handled when get_final_response is called again
        ensure
          stream_done = true
        end
      end

      begin
        # Yield partial results as they come in
        last_index = 0
        while !stream_done && !cancelled?
          # Yield any new partial results
          while last_index < partial_results.length
            yield partial_results[last_index]
            last_index += 1
          end
          
          # Small delay to avoid busy waiting
          sleep(0.01)
        end
        
        # Yield any remaining partial results
        while last_index < partial_results.length
          yield partial_results[last_index]
          last_index += 1
        end
        
      ensure
        # Clean up
        @ffi_stream.on_event(nil) if @ffi_stream.respond_to?(:on_event)
        stream_thread.join(1.0) # Give it a second to clean up
      end
    end

    private

    # Override these methods in subclasses to provide proper type coercion
    sig { params(result: T.untyped).returns(PartialType) }
    def coerce_partial(result)
      result
    end

    sig { params(result: T.untyped).returns(FinalType) }
    def coerce_final(result)
      result
    end
  end
end
    end

    # Calls the given block once for each event in the stream, where event is a parsed
    # partial response. Returns `self` to enable chaining `.get_final_response`.
    #
    # Must be called with a block.
    #
    # @yieldparam [PartialType] event the parsed partial response
    # @return [BamlStream] self
    sig { params(block: T.proc.params(event: PartialType).void).returns(BamlStream)}
    def each(&block)
      # Implementing this and include-ing Enumerable allows users to treat this as a Ruby
      # collection: https://ruby-doc.org/3.1.6/Enumerable.html#module-Enumerable-label-Usage
      if @final_response == nil
        @final_response = @ffi_stream.done(@ctx_manager) do |event|
          block.call event.parsed_using_types(Baml::Types, Baml::PartialTypes, true)
        end
      end

      self
    end


    # Gets the final response from the stream.
    #
    # @return [FinalType] the parsed final response
    sig {returns(FinalType)}
    def get_final_response
      if @final_response == nil
        @final_response = @ffi_stream.done(@ctx_manager)
      end

      @final_response.parsed_using_types(Baml::Types, Baml::PartialTypes, false)
    end
  end

end